c/src/lib/libbsp/arm/shared/stm32fxxxx/spi/hal-spi.c

```diff
@@ -0,0 +1,507 @@
+               break;
+   case SPI_THREE:   return (IRQn_Type)GET_SPI_VEC(SPI_THREE++);
+               break;
+   case SPI_FOUR:   return (IRQn_Type)GET_SPI_VEC(SPI_FOUR++);
+               break;
+   case SPI_FIVE:   return (IRQn_Type)GET_SPI_VEC(SPI_FIVE++);
+               break;
+   case SPI_SIX:   return (IRQn_Type)GET_SPI_VEC(SPI_SIX++);
+               break;
+   }
+}
+
+
+/**
+ * SPI IOCTL
+ */
+static int stm32_spi_ioctl(
+  spi_bus *base,
+  uint32_t ioctl_cmd,
+  unsigned long arg
+)
+{
+   rtems_status_code sc = RTEMS_SUCCESSFUL;
+   stm32_spi_bus *bus = (stm32_spi_bus *) base;
+
+   _Assert(base != NULL);
+
+   switch(ioctl_cmd)
+   {
+      /* Two line half-duplex mode */
+      case SPI_IOCTL_DIR_2LINE:    __HAL_SPI_ENABLE(bus->handle);
+                           bus->handle->Instance->CR1 &= ~(SPI_CR1_BIDIMODE);
+                           __HAL_SPI_DISABLE(bus->handle);
+                           return 0;
+                           break;
+
+      /* bi-direction line full-duplex mode */
+      case SPI_IOCTL_DIR_1LINE:   __HAL_SPI_ENABLE(bus->handle);
+                           bus->handle->Instance->CR1 |= SPI_CR1_BIDIMODE;
+                           __HAL_SPI_DISABLE(bus->handle);
+                           return 0;
+                           break;
+
+      /* Configure SPI instance to Master mode */
+      case SPI_IOCTL_MODE_MASTER: __HAL_SPI_DISABLE(bus->handle);
+                           bus->handle->Instance->CR1 |= (SPI_CR1_MSTR | SPI_CR1_SSI);
+                           __HAL_SPI_ENABLE(bus->handle);\
+                           return 0;
+                           break;
+
+      /* Configure SPI instance to Slave mode */
+      case SPI_IOCTL_MODE_SLAVE:   __HAL_SPI_DISABLE(bus->handle);
+                           bus->handle->Instance->CR1 &= ~(SPI_CR1_MSTR | SPI_CR1_SSI);
+                           __HAL_SPI_ENABLE(bus->handle);
+                           return 0;
+                           break;
+
+      /* Enable CRC Calculation */
+      case SPI_IOCTL_ENABLE_CRC:   __HAL_SPI_DISABLE(bus->handle);
+                           bus->handle->Instance->CR1 |= SPI_CR1_CRCEN;
+                           __HAL_SPI_ENABLE(bus->handle);
+                           return 0;
+                           break;
+      /* Disable CRC Calculation */
+      case SPI_IOCTL_DISABLE_CRC:   __HAL_SPI_DISABLE(bus->handle);
+                           bus->handle->Instance->CR1 &= ~(SPI_CR1_CRCEN);
+                           __HAL_SPI_DISABLE(bus->handle);
+                           return 0;
+                           break;
+
+      default:               return -1;
+   }
+}
+
+/**
+ * SPI Interrupt Event ISR
+ */
+static void stm32_spi_event_irq(void* arg)
+{
+   rtems_status_code sc;
+   stm32_spi_bus *bus;
+
+   HAL_SPI_IRQHandler((SPI_HandleTypeDef *) arg);
+
+   sc = rtems_event_transient_send(bus->task_id);
+   _Assert(sc == RTEMS_SUCCESSFUL);
```

Comment:
i don't think assert is doing what you want it to do most of the time, as per our earlier discussion. I believe for stuff like this if statements shoudl be used as _Assert can be disabled and it is not clear to me we want things like this to be disabled when not in debug mode.